### ISSUE TITLE
A bug to run h2ogpt-4096-llama2-70b-chat

### ISSUE THREAD

**Issue by babytdream** [h2oai/h2ogpt#692]
![image](https://github.com/h2oai/h2ogpt/assets/74184102/39a7ab49-6764-4b0b-87e6-2328b0d0d00e)
I have 16 * A10. But when I run this recommend **“python generate.py --base_model=meta-llama/Llama-2-70b-chat-hf --prompt_type=llama2 --rope_scaling="{'type': 'linear', 'factor': 4}" --use_gpu_id=False --save_dir=savemeta70b”**，it appears a BUG:
![image](https://github.com/h2oai/h2ogpt/assets/74184102/97bf2178-b21a-4ce7-910d-4cbfca721258)

**Comment by pseudotensor** [comment]
https://discuss.pytorch.org/t/cuda-error-peer-mapping-resources-exhausted/167814

**Comment by babytdream** [comment]
Hello，8 gpu is ok！But more gpu is helpful!
Do you use "torch.nn.DataParalle" or "torch.nn.parallel.DistributedDataParallel" in h2ogpt?Thanks！

**Comment by pseudotensor** [comment]
@arnocandel would know best about that.

**Comment by babytdream** [comment]
I find that this error only appears GPU > 11.
When  I use GPU <= 10, it can work! Like this command:**CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7,8,9 python generate.py --base_model=/data/model/h2ogpt-4096-llama2-70b-chat/ --prompt_type=llama2 --use_gpu_id=False --share=True**
![image](https://github.com/h2oai/h2ogpt/assets/74184102/41bc6d0c-72a7-4b5d-b82e-dc0decb540ff)

**Comment by pseudotensor** [comment]
Good!  @arnocandel would know best about how to run h2oGPT in distributed mode to do 8*8 mode

**Comment by arnocandel** [comment]
I haven't ever done distributed generation with > 8 GPUs, just distributed training (via `ddp=False`) but even that is currently not working, WIP to bring back: https://github.com/h2oai/h2ogpt/pull/644, but lower priority since H2O LLM Studio is working on DeepSpeed https://github.com/h2oai/h2o-llmstudio/pull/288

More GPUs doesn't make it faster, just allows for larger models:
https://github.com/h2oai/h2ogpt/blob/main/benchmarks/perf.md

**Comment by babytdream** [comment]
@arnocandel Thanks！The use of 16 GPUs is mainly to fine-tune the big model like LLama-70B using lora technology.Do you have some ideas to fine-tune using 16 GPUs?

**Comment by pseudotensor** [comment]
@babytdream For fine-tuning, best to try out LLM Studio.  It's more well-developed and easy to use for fine-tuning than h2oGPT: https://github.com/h2oai/h2o-llmstudio

**Comment by arnocandel** [comment]
@babytdream you can certainly fine-tune using 2 separate processes, each using 8 GPUs:

I did something similar here across 3 machines/processes:
https://github.com/h2oai/h2ogpt/blob/542f20495c46fcb44d60c05e1c8fde60e15f8aeb/finetune.py#L664-L669

You could try something like this:
```
NCCL_P2P_LEVEL=LOC WORLD_SIZE=16 CUDA_VISIBLE_DEVICES="0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15" torchrun --nproc_per_node=16 --master_port=1234 finetune.py --micro_batch_size=1 --batch_size=16
```

**Comment by babytdream** [comment]
@arnocandel OK！I see the data/config.json. If I try to fintune LLama-2-70B, should I modify the parameters like **"prompt_type"**?
      [
      {
          "prompt_type": **"llama2"**,
          "instruction": "Explain the following expert setting for Driverless AI",
          "input": "text.....",
          "output": "text..... "
      },
      ]

**Comment by arnocandel** [comment]
yeah, above command is assuming that the model fits on every GPU (and GPUs are used to train more mini batches at once), which the 70B doesn't, not even with 4bit.

**Comment by arnocandel** [comment]
#643 would be needed

**Comment by babytdream** [comment]
I also ask the same question in LLM Studio, they can't solve it.
@arnocandel sorry！Can you be more specific, it's a bit difficult for me. Or give me a similar command.
